### ISSUE TITLE
Diffusion Equation (moisture sorption in porous material)(FNN)

### ISSUE THREAD

**Issue by matth86** [lululxvi/deepxde#347]
Dear Prof. Lu Lu,
thank you for writing the deepxde library. I'm trying to solve this diffusion equation with PINN
**(short: Could anybody tell me if I coded the PDE (seen below)+IC/BC correctly?)**

![sorption](https://user-images.githubusercontent.com/34093797/126907540-2cb683e2-d4bb-4edb-a37a-3e9f947d79f1.png)
[https://drive.google.com/file/d/1hCRUe8WKlAk1zdjwk_TyJ6kMVprIjsxU/view?usp=sharing](url)

Well rather I'm trying to replicate the solution of an old paper
![dimlessQ](https://user-images.githubusercontent.com/34093797/126907568-ef22c6f4-1f1f-4f4e-94f3-1888493dae42.png)
Wanting to understand the procedures overall, as I will try to use it for a reverse problem later on with experimental data.

I think I'm doing sth wrong. I've looked at your examples but with my limited experience with python I'm having a hard time. Especially the setting the PDE and boundary conditions. All values (alpha,gamma,P,..) are taken from the paper (Ind. Eng. Chem. Res. 2000, 39, 219-226).

```python
import numpy as np

import matplotlib.pyplot as plt
import deepxde as dde
from deepxde.backend import tf

  
e=0.5#porosity
P=1.164E-4
alpha=3792
beta=12.6
gamma=55385
L=1#dimensionless length
T=1
Q_0=0.85
    
#setting up the geometry and the time  
geom = dde.geometry.Interval(0, L)
timedomain = dde.geometry.TimeDomain(0, T)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

#define the PDEs system
def pde(x, y):
  C, Q = y[:, 0:1], y[:, 1:2]
  dC_t = dde.grad.jacobian(y, x, i=0, j=1)
  dC_xx = dde.grad.hessian(y, x, component=0, i=0, j=0)
  dQ_t = dde.grad.jacobian(y, x, i=1, j=1)
  dQ_xx = dde.grad.hessian(y, x, component=1, i=0, j=0)      #not needed, should I still set this? Will be set automatically in .pde?

  eq_C = e*dC_t - 1/P * dC_xx + alpha * dQ_t 
  eq_Q = dQ_t - beta * (C - Q)# C instead of Qsat,bc linearization of Qsat
  return [eq_C, eq_Q]

#functions for initial 
#def fun_init(x):
#  return 0
    
#functions for bc 

def bcC_left(inputs, outputs, _):
    # 
    rs=gamma*P*(inputs[:,1:]-outputs[:, 0:1])
    # 
    dC = tf.gradients(outputs, inputs)[0]
    dC_dx = dC[:, 0:1]
    return dC_dx + rs
    
#def fun_bc_left(x):
#  return (-1)*gamma*P*x[:,1:]*y[:, 0:1]
    
def fun_bc_right(x):
  return 0

#functions to define which points belong to boundaries

def boundary_left(x, on_boundary):
  return on_boundary and np.isclose(x[0], 0)

def boundary_right(x, on_boundary):
  return on_boundary and np.isclose(x[0], L)
    

#define boundary conditions 
bc_Cleft = dde.OperatorBC(geomtime, bcC_left, boundary_left)   
bc_Cright = dde.NeumannBC(geomtime, fun_bc_right, boundary_right)

bc_Qleft = dde.DirichletBC(geomtime, lambda x: Q_0, boundary_left)
bc_Qright = dde.NeumannBC(geomtime, fun_bc_right, boundary_right)    
    
#define initial conditions   
    
ic1 = dde.IC(geomtime, lambda x: 0, lambda _, on_initial: on_initial, component=0)
ic2 = dde.IC(geomtime, lambda x: 0, lambda _, on_initial: on_initial, component=1)

#put all together into a data container to feed the network afterwards

data = dde.data.TimePDE(
  geomtime,
  pde,
  [ic1, ic2, bc_Cleft, bc_Cright,bc_Qleft, bc_Qright],
  num_domain=3000,
  num_boundary=300,
  num_initial=300,
  num_test=300,
)

#define the network    
net = dde.maps.FNN([2] + [50] * 5 + [2], "tanh", "Glorot uniform")
#net.apply_output_transform(lambda x, y: x[:, 0:1] + .85) hard boundary
model = dde.Model(data, net)

#early_stopping = dde.callbacks.EarlyStopping(min_delta=1e-4, patience=2000)
model.compile("adam", lr=1e-3)
#losshistory, train_state=model.train(epochs=25000, disregard_previous_best=True, callbacks=[early_stopping])
losshistory, train_state=model.train(epochs=25000, disregard_previous_best=True)

#save the output data
#model.compile("L-BFGS-B")
#losshistory, train_state = model.train()    
dde.saveplot(losshistory, train_state, issave=True, isplot=True)
```
Training Output:
![train](https://user-images.githubusercontent.com/34093797/126908286-8018a211-6ae6-4c0f-b8fd-8230cff0f6bb.png)

I will further look through your closed issues but any help would be appreciated. 
Thank you
Matth

**Comment by lululxvi** [comment]
The first issue of your code is the scaling. You can check FAQ for this.

By the way, you don't need `OperatorBC`. You can simply use `RobinBC`.

**Comment by matth86** [comment]
@lululxvi 
Hello. Thanks for your help. 

> The first issue of your code is the scaling. You can check FAQ for this.

Ok, scaling. I might misunderstood "Q: I failed to train the network or get the right solution, e.g., the training loss is large". I understood it as: it is only important that the output is between 0 and 1.

````python
eq_C = e*dC_t - 1/P * dC_xx + alpha * dQ_t 
````
in this case we want 
````python
1/P = 1
alpha =1
````
and then scaling the output by
net.apply_output_transform(lambda x, y: y[:,0:1]*1/P, y[:,1:]*alpha) ~these might be the wrong scaling factors

> You can also use hard constraints for some BC/IC.

**There I would have an additional question.** Starting at a basic diff-equation:
![startBasicDiff](https://user-images.githubusercontent.com/34093797/127730221-a3785eb8-f84a-45fd-bba8-9a94e800f82b.png)
I've tried out enforcing boundary conditions. 
![startBasicDiffBC](https://user-images.githubusercontent.com/34093797/127730225-516878f3-bdd7-4efe-8a34-e05ff0200e2d.png) (the IC looks different see in the code.. to agree with the BC)
I've run into a problem:
![left_hardbc](https://user-images.githubusercontent.com/34093797/127730232-81257f27-8d9a-4251-98d9-2702e8fc2f07.png)
The NN starts to output negative c* to agree with the BC and IC. Can I restrict the NN from outputting negative values? I though I can just use "sigmoid" than it should be always positive but my limited experience with NN proved me wrong. Further I tried transforming the output with sth like this
````python
net.apply_output_transform(lambda x, y: tf.cond(tf.math.less(tf.constant([0.]),y),lambda: y, lambda: tf.constant([0.])))
````
but this gets me an error.
Then I read #90
>lululxvi wrote: "In both cases, you solution is discontinuous, which could be hard to solve. `tf.sign` seems correctly, but maybe there is something wrong in computing the derivative in Tensorflow (I am not sure about this). You may only enforce one of them by `apply_output_transform`".

So there is no easy remedy for this case?

[https://colab.research.google.com/drive/1QjhSNBkzzn6BRz4cMYvOJo6s8czJBAvu?usp=sharing](url) (code on google colab)
````python
#1D Diffusion (dimless)
L=1
T=1
def pde(x, y):#y -> c
    #if tf.math.less(y,0.) is not None:
    #  y=0.
    dy_t = dde.grad.jacobian(y, x, i=0, j=1)
    dy_xx = dde.grad.hessian(y, x, i=0, j=0)
    return dy_t - dy_xx

geom = dde.geometry.Interval(0, L)
timedomain = dde.geometry.TimeDomain(0, T)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

def boundary_left(x, on_boundary):
  return on_boundary and np.isclose(x[0], 0)

def boundary_right(x, on_boundary):
  return on_boundary and np.isclose(x[0], L)

ic = dde.IC(
  geomtime, lambda x: np.exp(-x[:,0:1]*9999), lambda _, on_initial: on_initial
)

bc_l = dde.DirichletBC(geomtime, lambda x: 1, boundary_left)
bc_r = dde.DirichletBC(geomtime, lambda x: 0, boundary_right)

data = dde.data.TimePDE(
    geomtime,
    pde,
    [bc_l, bc_r, ic],
    num_domain=1000,
    num_boundary=1500,
    num_initial=3300
)
net = dde.maps.FNN([2] + [30]*3 + [1], "tanh", "Glorot normal")
net.apply_output_transform(lambda x, y: x[:, 0:1] * y+1)
#net.apply_output_transform(lambda x, y: tf.cond(tf.math.less(tf.constant([0.]),y),lambda: y, lambda: tf.constant([0.])))

model = dde.Model(data, net)
model.compile("adam", lr=1.0e-3,loss_weights=[1,10,1,10])
losshistory, train_state=model.train(epochs=30000, disregard_previous_best=True)


dde.saveplot(losshistory, train_state, issave=True, isplot=True)
````

> By the way, you don't need OperatorBC. You can simply use RobinBC.

Thanks for the hint. 

Thank you for your help
Matth

**Comment by lululxvi** [comment]
Your IC and left BC are not consistent. Modify your IC a little bit so that they are consistent, and then use hard constraints for both IC and BC.

**Comment by matth86** [comment]
Well I guess this should make the BC and IC consistent?
![newIC_BC](https://user-images.githubusercontent.com/34093797/128314737-d8c0a415-b4bb-4ba5-9b7e-972b4d26b92c.png)

````python3
ic = dde.IC(geomtime, lambda x: np.exp(-x[:,0:1]*99.), lambda _, on_initial: on_initial)
bc_l = dde.DirichletBC(geomtime, lambda x: 1., boundary_left)
bc_r = dde.DirichletBC(geomtime, lambda x: 0., boundary_right)
````
By reading through FAQ:"Q:By default, initial/boundary conditions are enforced in DeepXDE as soft constraints. How can I enforce them as hard constraints?"
I know I can use

[https://colab.research.google.com/drive/1QjhSNBkzzn6BRz4cMYvOJo6s8czJBAvu?usp=sharing](url)
````python3
def modify_output(X, y):
    x, t = X[:, 0:1], X[:, 1:2]
    c= y[:, 0:1]
    #c_new=t*c #set value of IC
    c_new = x*(1.-x)*c+1.-x #set values of BC
    #*tf.math.sign(t) not backward propagation compatible
    return tf.concat(c_new, axis=1)

...
net.apply_output_transform(modify_output)
````
for setting hard boundary conditions. Either I can set IC(x,t=0/1) a specific value or set the BC(x=0/1,t) a specific value. 

(e.g. forcing BC with reduced timedomain)
![leftright_hardbc](https://user-images.githubusercontent.com/34093797/128331186-48cc87ce-ceaf-491d-86c9-53bfefeac58f.png)
`and then use hard constraints for both IC and BC`
What I do not understand: **how can I enforce my IC, exp(-99x) only for t=0?** 

Do I miss something obvious?
Thank you
Matth

**Comment by lululxvi** [comment]
You are close.
```python
def modify_output(X, y):
    x, t = X[:, 0:1], X[:, 1:2]
    c = y
    c_new = t*x*(1-x)*c + tf.exp(-x * 99.)
    return c_new
```

**Comment by matth86** [comment]
Thanks for your help @lululxvi

I think I am pushing the PINN over the edge by forcing this IC. 
I would assume the backward propagation does not like the exp(...).

![IC-problem](https://user-images.githubusercontent.com/34093797/128865072-2c760760-095b-442a-85bc-afe34c031a85.png)

I will try to set different hyperparameters.

**Comment by matth86** [comment]
It is not the exp(...) but the coefficient a in exp(-ax). If a gets to large it seems to find a very wrong local minima.
I've compared different cases with an explicit finite differencing scheme (below is a case where only the boundary is forced but the ic is soft) over the geomtime domain:
![absrelerr# h_1_30_3_100000_n_referencevs9999](https://user-images.githubusercontent.com/34093797/129490253-0d862d48-b4a4-4bcc-98f5-6561b3c4ab3d.png)
near the IC the prediction has the most deviation from the FD method. 
Dear Prof. @lululxvi 
**Anyway I'm trying now to use DEEPXDE for the same problem but inverse (dimless 2nd Fick law) but the results are not near the expected value D=1**
````python3
!pip install deepxde
import re
import numpy as np
import matplotlib.pyplot as plt
import deepxde as dde
from deepxde.backend import tf
from google.colab import files

L=1.
T=1.
header="var05.dat"

def gen_traindata():
    data=np.genfromtxt('data_251U101.dat',delimiter=' ') 
    x=data[:,0:1] 
    y=data[:,1:2] 
    z=data[:,2:3] 
    return np.hstack((x, y)), z

D = tf.Variable(1.0)

def pde(x, y):
    dy_t = dde.grad.jacobian(y, x, i=0, j=1)
    dy_xx = dde.grad.hessian(y, x, i=0, j=0)
    return dy_t - D*dy_xx

geom = dde.geometry.Interval(0., L)
timedomain = dde.geometry.TimeDomain(0., T)
geomtime = dde.geometry.GeometryXTime(geom, timedomain)

def boundary_left(x, on_boundary):
  return on_boundary and np.isclose(x[0], 0.)

def boundary_right(x, on_boundary):
  return on_boundary and np.isclose(x[0], L)

ic = dde.IC(
  geomtime, lambda x: np.exp(-x[:,0:1]*9999.), lambda _, on_initial: on_initial
)

bc_l = dde.DirichletBC(geomtime, lambda x: 1., boundary_left)
bc_r = dde.DirichletBC(geomtime, lambda x: 0., boundary_right)

obs_X, c = gen_traindata()
obs_c = dde.PointSetBC(obs_X, c, component=0)


data = dde.data.TimePDE(
    geomtime,
    pde,
    [bc_l, bc_r, ic, obs_c],
    num_domain=1,
    num_boundary=2,
    num_initial=1,
    anchors=obs_X,
    #num_test=50000,
)
net = dde.maps.FNN([2] + [30] * 3 + [1], "tanh", "Glorot uniform")

model = dde.Model(data, net)
model.compile("adam", lr=0.001, external_trainable_variables=D)
variable = dde.callbacks.VariableValue(D, period=1000, filename=header)
losshistory, train_state = model.train(epochs=80000, callbacks=[variable])
dde.saveplot(losshistory, train_state, issave=True, isplot=True)

files.download(header) 

lines = open(header, "r").readlines()
# read output data in fnamevar (this line is a long story...)
Chat = np.array(
[
  np.fromstring(
  min(re.findall(re.escape("[") + "(.*?)" + re.escape("]"), line), key=len),
  sep=",",
  )
        for line in lines
]
)
l, c = Chat.shape

plt.plot(range(l), Chat[:, 0], "r-")
plt.plot(range(l), np.ones(Chat[:, 0].shape), "r--")
plt.legend(["D","True D"], loc="right")
plt.xlabel("Epoch")
plt.ylabel("D")
plt.show()
````
I tried using random points over the whole domain, uniform points,.... 
![small1](https://user-images.githubusercontent.com/34093797/129490517-2e7a976f-02e7-4839-a9e5-0eb8a4f73c3d.png)
![small2](https://user-images.githubusercontent.com/34093797/129490529-ba4b0c4a-a7e5-44b1-98da-b695fada1151.png)
![small3](https://user-images.githubusercontent.com/34093797/129491138-b18a33d3-ded4-4132-8b29-03f6e28622f7.png)

but the results change between 0.6-2.27
80000 [8.57e-01] with softboundary, uniform ~25000 points (here domain training points = 1, as seen in the code)
80000 [2.74e+00] with softboundary, random 10000 points (here domain training points = 2000)
76000 [6.22e-01] with hardboundary, random 10000 points (here domain training points = 2000)
....
I guess my FD scheme could be wrong(I dont think so) . I will further read the FAQ and the issues on the same topic and further study all the examples you have already given. **I just wanted to ask if you can see an obvious error?**
Thank you 
Matth

**Comment by lululxvi** [comment]
Is `D` trainalbe? If not, then use `D=1.0`, not `D = tf.Variable(1.0)`

**Comment by matth86** [comment]
Thanks Prof. @lululxvi 

Changing the IC would be my next plan. 

I used your [https://github.com/lululxvi/deepxde/blob/95ad89a0678ae6dbeb5d87c1c42f5d3f07bf08b5/examples/reaction_inverse.py](test)
inverse reaction case to study the behavior of a PINN in finding k and D with different hyperparameters.

With this , I could get it at least near the expected result D=1 but I'm still tinkering around with it.
![predVAR](https://user-images.githubusercontent.com/34093797/130016268-3eec99e3-25b5-45ac-92f9-6cfa41c4e712.png)
(image is showing the same case(same trainingpoints, bcs,...), just the epochs are different. i.e. ir06=80k, ir07=160k,...)
